File: unwordle.py
#!/usr/bin/python3 -u
from wordle import word_delta, word_list
from pathlib import Path
from sys import stdin, stdout, stderr
from argparse import ArgumentParser
from statistics import mean

def freq(data):
	return (sum(map(lambda x: x==y, data)) for y in set(data))

def max_group_size(word, word_list):
	return max(freq(map(lambda w: word_delta(word, w), word_list)))

def mean_group_size(word, word_list):
	return mean(freq(list(map(lambda w: word_delta(word, w), word_list))))

def negative_count_unique_results(word, word_list):
	return -len(set(map(lambda w: word_delta(word, w), word_list)))

def debug(*args, file=stderr, **kwargs):
	pass
#	return print(*args, file=file, **kwargs)


def unwordle(
    word_list, 
    guess, 
    score=negative_max_group_size, 
    input=stdin, 
    output=stdout):
	debug(score)
	def try_word(word):
		print(word, file=output)
		output.flush()

	def pick_word(candidates):
		return min(candidates, key=lambda w: score(w, candidates))

	if guess is None:
		guess = pick_word(word_list)

	try_word(guess)

	guesses=1
	candidates = word_list.copy()

	for result in input:
		candidates = list(filter(lambda w: word_delta(guess, w) == result.strip(), candidates))
		debug(candidates if len(candidates) < 10 else len(candidates))
		if not candidates:
			try_word("tears")
			return ("failed", -guesses)
		elif len(candidates) == 1:
			guess = candidates[0]
		else:
			guess = pick_word(candidates)
		try_word(guess)
		guesses += 1
	output.close()

	return (guess, guesses)

if __name__ == '__main__':
	try:
		import signal
		signal.signal(signal.SIGPIPE, signal.SIG_DFL)
	except (ImportError, AttributeError):
		# Do nothing on platforms without signals or ``SIGPIPE``.
		pass

	parser = ArgumentParser()
	parser.add_argument('-g', '--guess', dest="guess")
	parser.add_argument('-f', '--word-list', dest="word_list", default='wordles.txt', type=word_list)
	args = parser.parse_args()
	print(*unwordle(**vars(args)), file=stderr)

File: unwordle.jl
#!/usr/bin/env -S julia --threads auto --project --startup-file=no 
#-e 'using DaemonMode; runargs()'
using Statistics
using ArgParse
using DataStructures: counter
using ThreadsX

word_diff(A, B) = sum(3^i * if a==b; 1 elseif a in B; 2 else 3 end for (i, (a,b)) in enumerate(zip(A, B)))

word_diff1(A,B) = join(if a==b; a; elseif a in B; '+' else '-' end for (a,b) in zip(A,B))

parse_diff(s) = sum(3^i * if a=='-'; 3 elseif a=='+'; 2; else 1 end for (i, a) in enumerate(s))

const tmap = ThreadsX.map
group_sizes(word, words) = values(counter(tmap(identity, (word_diff(word,w) for w in words))))

entropy(word, words) = let W = length(words); sum(S/W * log(S/W) for S in group_sizes(word,words)) ; end

function unwordle(
	word_list, 
	guess=nothing;
	score=entropy, 
	input=stdin, 
	output=stdout
	)

	function try_word(word) 
		println(output, word)
		flush(output)
	end

	function pick_word(words) 
		(word_score, index) = findmin(w-> score(w, words), word_list)
		@debug "" index word_list[index]
		return word_list[index]
	end
	
	if guess == nothing
		guess = pick_word(word_list)
	end

	try_word(guess)
	guesses = 1

	candidates = copy(word_list)
	@debug " " length(candidates) guesses

	for result in eachline(input)
		filter!(w-> word_diff1(guess, w) == result, candidates)
		@debug " " guesses ifelse(length(candidates) < 10, candidates,length(candidates))

		if isempty(candidates)
			error("ran out of words to guess from")
		elseif length(candidates) == 1
			guess = only(candidates)
		else
			guess = pick_word(candidates)
		end
		try_word(guess)
		guesses += 1
	end
	return (word=guess, guesses=guesses)
end


function main()
	parser = ArgParseSettings()
	@add_arg_table parser begin
		"--guess", "-g"
			help = "the initial guess"
			default = nothing
			arg_type = Union{Nothing, String}
		 "--word-list", "-f"
		help="word list text file"
		default = "./wordles.txt"
	end
	args = parse_args(parser)
	result = unwordle(readlines(args["word-list"]), args["guess"])
	join(stderr, result, " ")
	println(stderr)
end


if abspath(PROGRAM_FILE) == @__FILE__
    # do something only this file is executed. 
	main()
end
